sql/src/main/java/io/crate/action/sql/DDLStatementDispatcher.java

```diff
@@ -103,22 +105,32 @@ public DDLStatementDispatcher(Provider<BlobAdminClient> blobAdminClient,
 
         @Override
         public CompletableFuture<Long> visitOptimizeTableStatement(OptimizeTableAnalyzedStatement analysis, UUID jobId) {
-            ForceMergeRequest request = new ForceMergeRequest(analysis.indexNames().toArray(new String[0]));
-
-            // Pass parameters to ES request
-            request.maxNumSegments(analysis.settings().getAsInt(OptimizeSettings.MAX_NUM_SEGMENTS.name(),
-                ForceMergeRequest.Defaults.MAX_NUM_SEGMENTS));
-            request.onlyExpungeDeletes(analysis.settings().getAsBoolean(OptimizeSettings.ONLY_EXPUNGE_DELETES.name(),
-                ForceMergeRequest.Defaults.ONLY_EXPUNGE_DELETES));
-            request.flush(analysis.settings().getAsBoolean(OptimizeSettings.FLUSH.name(),
-                ForceMergeRequest.Defaults.FLUSH));
-
-            request.indicesOptions(IndicesOptions.lenientExpandOpen());
-
-            FutureActionListener<ForceMergeResponse, Long> listener =
-                new FutureActionListener<>(Functions.constant((long) analysis.indexNames().size()));
-            transportActionProvider.transportForceMergeAction().execute(request, listener);
-            return listener;
+            if (analysis.settings().getAsBoolean(OptimizeSettings.UPGRADE_SEGMENTS.name(),
+                OptimizeSettings.UPGRADE_SEGMENTS.defaultValue())) {
+                UpgradeRequest request = new UpgradeRequest(analysis.indexNames().toArray(new String[0]));
+
+                FutureActionListener<UpgradeResponse, Long> listener =
+                    new FutureActionListener<>(Functions.constant((long) analysis.indexNames().size()));
+                transportActionProvider.transportUpgradeAction().execute(request, listener);
+                return listener;
+            } else {
+                ForceMergeRequest request = new ForceMergeRequest(analysis.indexNames().toArray(new String[0]));
+
+                // Pass parameters to ES request
+                request.maxNumSegments(analysis.settings().getAsInt(OptimizeSettings.MAX_NUM_SEGMENTS.name(),
+                    ForceMergeRequest.Defaults.MAX_NUM_SEGMENTS));
+                request.onlyExpungeDeletes(analysis.settings().getAsBoolean(OptimizeSettings.ONLY_EXPUNGE_DELETES.name(),
+                    ForceMergeRequest.Defaults.ONLY_EXPUNGE_DELETES));
+                request.flush(analysis.settings().getAsBoolean(OptimizeSettings.FLUSH.name(),
+                    ForceMergeRequest.Defaults.FLUSH));
+
+                request.indicesOptions(IndicesOptions.lenientExpandOpen());
+
+                FutureActionListener<ForceMergeResponse, Long> listener =
+                    new FutureActionListener<>(Functions.constant((long) analysis.indexNames().size()));
+                transportActionProvider.transportForceMergeAction().execute(request, listener);
+                return listener;
+            }
```

Comment:
maybe it makes sense to move both to concrete methods (executeMergeSegments & executeUpgradeSegments)